### PR TITLE
🐛 make `listObjByGVK` resilient

### LINKED ISSUE #6261: Make listObjByGVK retry with backoff
**What steps did you take and what happened:**
the func `listObjByGVK` is used in clusterctl during the upgrade process, however this function does not implements a retry/backoff loop so it makes this bit of code sensible to transient errors that can happen during upgrades / in environments with poor network connections.

e.g.
```
Performing upgrade...
Scaling down Provider="cluster-api" Version="" Namespace="capi-system"
Scaling down Provider="bootstrap-kubeadm" Version="" Namespace="capi-kubeadm-bootstrap-system"
Scaling down Provider="control-plane-kubeadm" Version="" Namespace="capi-kubeadm-control-plane-system"
Scaling down Provider="infrastructure-aws" Version="" Namespace="capa-system"
Deleting Provider="cluster-api" Version="" Namespace="capi-system"
Error: failed to upgrade management cluster providers: failed to apply providers upgrade: failed to list objects for the "infrastructure.cluster.x-k8s.io/v1beta1, Kind=AWSClusterControllerIdentity" GroupVersionKind: conversion webhook for infrastructure.cluster.x-k8s.io/v1alpha3, Kind=AWSClusterControllerIdentity failed: Post "[https://capa-webhook-service.capa-system.svc:443/convert?timeout=30s](https://capa-webhook-service.capa-system.svc/convert?timeout=30s)": dial tcp 100.64.252.115:443: connect: connection refused
```

**What did you expect to happen:**
clusterctl to be resilient to temporary network problems.

**Anything else you would like to add:**
An example of retry look in clusterctl

https://github.com/kubernetes-sigs/cluster-api/blob/dcc22db766eb36c5548970fda3251efd2ebd595b/cmd/clusterctl/client/cluster/proxy.go#L237-L244

/kind bug

### PR DESCRIPTION
**Pull request by ykakarap**
**What this PR does / why we need it**:

This PR makes the `listObjByGVK` function resilient to temporary network problems. This function is used during the upgrade flow and the upgrade can potentially break because of this.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6261

### CODE REVIEW

**Review by killianmuldoon — APPROVED** [review]
/lgtm

**Comment by ykakarap** [comment]
/hold

To see if we it is okay to  address the first point mentioned here in the same PR: https://github.com/kubernetes-sigs/cluster-api/issues/6261#issuecomment-1060620932

File: cmd/clusterctl/client/cluster/proxy.go [hunk 1]
```diff
@@ -359,11 +359,15 @@ func listObjByGVK(c client.Client, groupVersion, kind string, options []client.L
...
 	objList.SetKind(kind)
 
-	if err := c.List(ctx, objList, options...); err != nil {
+	resourceListBackoff := newReadBackoff()
+	if err := retryWithExponentialBackoff(resourceListBackoff, func() error {
+		return c.List(ctx, objList, options...)
+	}); err != nil {
 		if !apierrors.IsNotFound(err) {
 			return nil, errors.Wrapf(err, "failed to list objects for the %q GroupVersionKind", objList.GroupVersionKind())
 		}
```
Comment (sbueringer) [inline cmd/clusterctl/client/cluster/proxy.go:368]: Q: will this error still be detected when the client-go is not found error is (probably?) wrapped in some kind of retry error?


Should this if be rather inside the retry anyway? (i.e. do we want to retry if the list call works but we simply can't find a resource)

Comment (fabriziopandini) [reply cmd/clusterctl/client/cluster/proxy.go:368]: Might be I'm wrong, but List should never return `apierrors.IsNotFound`, right?

Comment (sbueringer) [reply cmd/clusterctl/client/cluster/proxy.go:368]: Would make sense, I also don't know.
We should find out :)

Comment (fabriziopandini) [reply cmd/clusterctl/client/cluster/proxy.go:368]: I think you can entirely drop `if !apierrors.IsNotFound(err) { ` and then we can get this merged

**Pull request merged**